application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java

```diff
@@ -0,0 +1,147 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.togetherjava.tjbot.commands.SlashCommandAdapter;
+import org.togetherjava.tjbot.commands.SlashCommandVisibility;
+
+import java.util.Objects;
+
+/**
+ * This command can ban users and optionally remove their messages from the past days. Banning can
+ * also be paired with a ban reason. The command will also try to DM the user to inform them about
+ * the action and the reason.
+ * <p>
+ * The command fails if the user triggering it is lacking permissions to either ban other users or
+ * to ban the specific given user (for example a moderator attempting to ban an admin).
+ */
+public final class BanCommand extends SlashCommandAdapter {
+    private static final String USER_OPTION = "user";
+    private static final String DELETE_HISTORY_OPTION = "delete-history";
+    private static final String REASON_OPTION = "reason";
+    private static final Logger logger = LoggerFactory.getLogger(BanCommand.class);
+
+    /**
+     * Creates an instance of the ban command.
+     *
+     * @see BanCommand
+     */
+    public BanCommand() {
+        super("ban", "Bans the given user from the server", SlashCommandVisibility.GUILD);
+
+        getData().addOption(OptionType.USER, USER_OPTION, "The user who you want to ban", true)
+            .addOption(OptionType.STRING, REASON_OPTION, "Why the user should be banned", true)
+            .addOptions(new OptionData(OptionType.INTEGER, DELETE_HISTORY_OPTION,
+                    "the amount of days of the message history to delete, none means no messages are deleted.",
+                    true).addChoice("none", 0).addChoice("recent", 1).addChoice("all", 7));
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target is null");
+
+        Member target = userOption.getAsMember();
+
+        Member author = Objects.requireNonNull(event.getMember(), "The author is null");
+
+        String reason = Objects.requireNonNull(event.getOption(REASON_OPTION), "The reason is null")
+            .getAsString();
+
+        Guild guild = Objects.requireNonNull(event.getGuild());
+        Member bot = guild.getSelfMember();
+
+        // Member doesn't exist if attempting to ban a user who is not part of the guild.
+        if (target != null && !handleCanInteractWithTarget(target, bot, author, event)) {
+            return;
+        }
+
+        if (!handleHasPermissions(author, bot, event, guild)) {
+            return;
+        }
+
+        int deleteHistoryDays = Math
+            .toIntExact(Objects.requireNonNull(event.getOption(DELETE_HISTORY_OPTION)).getAsLong());
+
+        if (!ModerationUtils.handleReason(reason, event)) {
+            return;
+        }
+
+        banUser(userOption.getAsUser(), author, reason, deleteHistoryDays, event.getGuild(), event);
```

Comment:
You are still using `event.getGuild()` here, even though you introduced a `guild` variable beforehand...